pennylane/transforms/optimization.py

```diff
@@ -0,0 +1,254 @@
+            qml.RY(y, wires=1)
+            qml.PauliX(wires=1)
+            qml.RZ(z, wires=0)
+            qml.RX(y, wires=2)
+            qml.CNOT(wires=[0, 2])
+            qml.PauliX(wires=1)
+            return qml.expval(qml.PauliZ(0))
+
+    The circuit before optimization:
+
+    >>> dev = qml.device('default.qubit', wires=3)
+    >>> qnode = qml.QNode(qfunc, dev)
+    >>> print(qml.draw(qnode)(1, 2, 3))
+    0: ──H──────H──────RZ(3)─────╭C──┤ ⟨Z⟩
+    1: ──H──────RY(2)──X──────X──│───┤
+    2: ──RX(1)──RX(2)────────────╰X──┤
+
+
+    We can see that there are two adjacent Hadamards on the first qubit that
+    should cancel each other out. Similarly, there are two Pauli X gates on the
+    second qubit that should cancel. We can obtain a simplified circuit by running
+    the ``cancel_inverses`` transform:
+
+    >>> optimized_qfunc = cancel_inverses(qfunc)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+    >>> print(qml.draw(optimized_qnode)(1, 2, 3))
+    0: ──RZ(3)─────────╭C──┤ ⟨Z⟩
+    1: ──H──────RY(2)──│───┤
+    2: ──RX(1)──RX(2)──╰X──┤
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+
+        # Normally queue any gates that are not their own inverse
+        if current_gate.name not in self_inverse_symmetric:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # If a gate does has a self-inverse, find the next gate that acts on the same wires
+        next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If no such gate is found (either there simply is none, or there are other gates
+        # "in the way", queue the operation and move on
+        if next_gate_idx is None:
+            current_gate.queue()
+            list_copy.pop(0)
+            continue
+
+        # Otherwise, get the next gate
+        next_gate = list_copy[next_gate_idx + 1]
+
+        # If next gate is the same (self inverse), we can potentially remove it
+        if current_gate.name == next_gate.name:
+            # If the wires are the same, then we can safely remove
+            if current_gate.wires == next_gate.wires:
+                list_copy.pop(next_gate_idx + 1)
+            # If wires are not equal, need to check if the inverse is asymmetric;
+            # if it is not, then we can't cancel and have to queue it
+            else:
+                if current_gate.name in self_inverse_asymmetric:
+                    list_copy.pop(next_gate_idx + 1)
+                else:
+                    current_gate.queue()
+        # Otherwise, queue and move on to the next item
+        else:
+            current_gate.queue()
+
+        # Remove this gate from the working list
+        list_copy.pop(0)
+
+    # Queue the measurements normally
+    for m in tape.measurements:
+        m.queue()
+
+
+@qfunc_transform
+def merge_rotations(tape):
```

Comment:
This could even be combined into a single transform with inverse cancellation; I've separated them for now for easier development.

Reply:
I guess every transform that you combine, that is one fewer for loop 🤔 

Perhaps at some point it makes sense to be able to register transforms at the iteration level, and then allow them all to be composed at once into a single for loop.